File: src/main/java/com/rapidprototyping/manbot/controller/CommandHandler.java
package com.rapidprototyping.manbot.controller;

import com.rapidprototyping.manbot.model.command.ArgumentLength;
import com.rapidprototyping.manbot.model.user.UserType;

/**
 *
 * @author Pavan C (pavan407)
 */
public interface CommandHandler
{
    String getName();

    void handle(CommandContext ctx);

    default UserType getRequiredUserType()
    {
        return UserType.LEARNER;
    }

    default ArgumentLength getRequiredArgumentLength()
    {
        return ArgumentLength.anyLength();
    }

    default String getUsage()
    {
        return "N/A";
    }

    default String getAuthor()
    {
        return "Anonymous";
    }

    default String getDescription()
    {
        return "N/A";
    }
}

File: bot/src/main/kotlin/com/manbot/command/CommandHandler.kt
package com.manbot.command

import com.manbot.plugin.Plugin
import com.manbot.user.UserType

/**
 * @author Pavan C. (pavan407)
 */
abstract class CommandHandler @JvmOverloads constructor(
        val name: String,
        val description: String = "N/A",
        val argumentAmount: (Int) -> Boolean = ArgumentAmount.anyLength(),
        val requiredUserType: UserType = UserType.LEARNER,
        val usage: String = "N/A"
) : Plugin
{
    override fun onInit() = CommandHandlerCoordinator.applyCoordination(this)

    abstract fun handle(event: CommandEvent)
}